go/adbc/driver/snowflake/connection.go

```diff
@@ -557,13 +587,204 @@ func (c *cnxn) populateMetadata(ctx context.Context, depth adbc.ObjectDepth, cat
+			}
+		}
+	}
+
+	for key, constraintSchema := range fullyQualifiedConstraintSchemaLookup {
+		catalogSchemaTable := getCatalogSchemaTableFromFullyQualifiedConstraintName(key)
+		constraintLookup[catalogSchemaTable] = append(constraintLookup[catalogSchemaTable], constraintSchema)
+	}
+
+	return constraintLookup, nil
+}
+
+func (c *cnxn) getConstraintsData(ctx context.Context, depth adbc.ObjectDepth, metadataRecords []internal.Metadata) ([]TableConstraint, error) {
+	if depth == adbc.ObjectDepthCatalogs || depth == adbc.ObjectDepthDBSchemas {
+		return nil, nil
+	}
+	availableConstraintTypes := getAvailableConstraintTypes(metadataRecords)
+	availableFullyQualifiedConstraints := getAvailableFullyQualifiedConstraints(metadataRecords)
+
+	var uniqueConstraintsData []TableConstraint
+	var primaryKeyConstraintsData []TableConstraint
+	var foreignKeyConstraintsData []TableConstraint
+	var err error
+
+	if availableConstraintTypes != nil {
+		if _, exists := availableConstraintTypes[internal.Unique]; exists {
+			uniqueConstraintsData, err = c.getUniqueConstraints(ctx, availableFullyQualifiedConstraints)
+			if err != nil {
+				return nil, errToAdbcErr(adbc.StatusIO, err)
+			}
+		}
+
+		if _, exists := availableConstraintTypes[internal.PrimaryKey]; exists {
+			primaryKeyConstraintsData, err = c.getPrimaryKeyConstraints(ctx, availableFullyQualifiedConstraints)
+			if err != nil {
+				return nil, errToAdbcErr(adbc.StatusIO, err)
+			}
+		}
+
+		if _, exists := availableConstraintTypes[internal.ForeignKey]; exists {
+			foreignKeyConstraintsData, err = c.getForeignKeyConstraints(ctx, availableFullyQualifiedConstraints)
+			if err != nil {
+				return nil, errToAdbcErr(adbc.StatusIO, err)
+			}
+		}
+	}
+
+	tableConstraintsData := append(append(uniqueConstraintsData, primaryKeyConstraintsData...), foreignKeyConstraintsData...)
+
+	sort.Slice(tableConstraintsData, func(i, j int) bool {
+		if tableConstraintsData[i].constraintName == tableConstraintsData[j].constraintName {
+			return tableConstraintsData[i].keySequence < tableConstraintsData[j].keySequence
+		}
+		return tableConstraintsData[i].constraintName == tableConstraintsData[j].constraintName
+	})
+
+	return tableConstraintsData, nil
+}
+
+func (c *cnxn) getUniqueConstraints(ctx context.Context, fullyQualifiedConstraints map[string]bool) ([]TableConstraint, error) {
+	uniqueConstraintsData := make([]TableConstraint, 0)
+
+	rows, err := c.sqldb.QueryContext(ctx, prepareUniqueConstraintSQL(), nil)
+	if err != nil {
+		return nil, errToAdbcErr(adbc.StatusIO, err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var uniqueConstraint TableConstraint
+		if err := rows.Scan(&uniqueConstraint.skipCreatedOn, &uniqueConstraint.dbName, &uniqueConstraint.schema, &uniqueConstraint.tblName, &uniqueConstraint.colName, &uniqueConstraint.keySequence, &uniqueConstraint.constraintName, &uniqueConstraint.skipRely, &uniqueConstraint.skipComment); err != nil {
```

Comment:
split this line to make it easier to read

Reply:
split into multiple lines